docs/design/storagenode-downtime-tracking.md

```diff
@@ -0,0 +1,207 @@
+# Storage Node Downtime Tracking
+
+## Abstract
+
+The goal is tracking Storage Nodes downtime per Satellite.
+
+## Background
+
+The [Disqualification design document](disqualification.md) describes when the Storage Nodes get disqualified based on the reputation scores described in the [Node Selection design document](node-selection.md).
+
+The current disqualification showed to us that several Storage Nodes got disqualified because of their uptime stats without giving clear and fair evidence of it to us nor their operators. We had to take action and invalidate the uptime's part of the formula used for calculating their reputation.
+
+We need to enable back uptime disqualification, however before, we need to track in a fair way the amount of time Storage Nodes are offline.
+
+This document presents a solution for tracking  Storage Node offline time, but it's out of the scope of this document how Storage Node reputation is affected by it.
+
+The presented solution considers the requirements imposed by the [Kademlia removal design document](kademia-removal.md), where the Storage Nodes must contact the Satellite when they start and after, every hour.
+
+## Design
+
+For tracking Storage Nodes offline time we need:
+
+- A new SQL database table to store offline time.
+- Detecting which ones are offline.
+- Roughly estimate how long they are offline.
+
+An _uptime check_ referenced in this section is the connection initiated by the Satellite to any Storage Node in the same way that's described in [Network refreshing section of the Kademlia removal design document](kademia-removal.md#network-refreshing).
+
+__NOTE__ the SQL code in this section is illustrative for explaining the algorithm concisely.
+
+### Database
+
+The following new SQL database table will store Storage Nodes offline time.
+
+```sql
+CREATE TABLE nodes_offline_time(
+    node_id BYTEA NOT NULL,
+    tracked_at timestamp with time zone NOT NULL,
+    seconds integer,                                                    -- Measured number of seconds being offline
+    CONSTRAINT nodes_offline_time__pk PRIMARY KEY(node_id, tracked_at)
+)
+```
+
+The current Satellite database has the table `nodes`. For the offline time calculation we use the following columns:
+
+- `last_contact_success`
+- `last_contact_failure`
+- `total_uptime_count`
+- `uptime_success_count`
+
+### Detecting offline nodes
+
+An independent process runs, in a configurable interval, of time the following query:
+
+```sql
```

Comment:
so this process is querying for nodes that have been contacted over an hour ago, but where `last_contact_success > last_contact_failure`? Won't that give us mostly online nodes? I'd think we would want to check `last_contact_success < last_contact_failure` in this query in order to get offline nodes that have not been seen in the past hour.

Reply:
> so this process is querying for nodes that have been contacted over an hour ago, but where last_contact_success > last_contact_failure? Won't that give us mostly online nodes?

No, per Kademlia removal design document, the SNs have to ping the satellite every hour, hence if they don't do so, then we have to find out what happens:

1. SN did the ping but due "some network failure" the ping didn't reach the Satellite.
1. SN did the ping but the Satellite wasn't available.
1. SN didn't do the ping.

All of those 3 scenarios cause that the `last_contact_success` doesn't be updated in the 1-hour interval.

From the Satellite point of view, we don't know that of those 3 general cases could be so the way is doing a specific check; that's what this part of the algorithm does.

The Kademlia removal design document doesn't specify any mechanism for ensuring that SNs have contacted the Satellite every hour; this is the part that does such verification.

If the Satellite contacts before the 1-hour the SN (this happens for example by the audit process), then we have:

1. A proof that's online before the 1-hour interval. We can wait for one hour to hear from it again.
1. A proof that's offline. We have to start to measure its offline time, that's handled by the second part of the algorithm (section _Estimating offline time_).

There is another moment when SNs must ping the satellite when they startup (first time, reboot, etc.), but those SNs will be handled as the first case above.

> I'd think we would want to check last_contact_success < last_contact_failure in this query in order to get offline nodes that have not been seen in the past hour.

No, that's handled by the second part of the algorithm (section _Estimating offline time_).

That situation happens:

1. An SN hasn't contacted the Satellite in the 1-hour interval. The Satellite check the SN and it's offline and it updates the `last_contact_failure`. This is done by this part of the algorithm.
1. Another process of the satellite (e.g. Audit) has tried to contact the SN and it was offline. It updates the `last_contact_failure.